collector/corosync/parser.go

```diff
@@ -0,0 +1,213 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "could not parse node id in corosync-quorumtool output")
+	}
+
+	status.RingId, status.Seq, err = parseRingIdAndSeq(quorumToolOutput)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not parse ring id and seq number in corosync-quorumtool output")
+	}
+
+	status.Quorate, err = parseQuorate(quorumToolOutput)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not parse quorate in corosync-quorumtool output")
+	}
+
+	status.QuorumVotes, err = parseQuoromVotes(quorumToolOutput)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not parse quorum votes in corosync-quorumtool output")
+	}
+
+	status.Members, err = parseMembers(quorumToolOutput)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not parse members in corosync-quorumtool output")
+	}
+
+	status.Rings = parseRings(cfgToolOutput)
+
+	return status, nil
+}
+
+func parseNodeId(cfgToolOutput []byte) (string, error) {
+	nodeRe := regexp.MustCompile(`(?m)Node ID:\s+(\w+)`)
+	matches := nodeRe.FindSubmatch(cfgToolOutput)
+	if matches == nil {
+		return "", errors.New("could not find Node ID line")
+	}
+
+	return string(matches[1]), nil
+}
+
+func parseRingIdAndSeq(cfgToolOutput []byte) (string, uint64, error) {
+	nodeRe := regexp.MustCompile(`(?m)Ring ID:\s+(\w+)/(\d+)`)
+	matches := nodeRe.FindSubmatch(cfgToolOutput)
+	if matches == nil {
+		return "", 0, errors.New("could not find Ring ID line")
+	}
+
+	seq, err := strconv.Atoi(string(matches[2]))
+	if err != nil {
+		return "", 0, errors.Wrap(err, "could not parse seq number to uint64")
+	}
+
+	return string(matches[1]), uint64(seq), nil
+}
+
+func parseQuorate(quorumToolOutput []byte) (bool, error) {
+	re := regexp.MustCompile(`(?m)Quorate:\s+(Yes|No)`)
+	matches := re.FindSubmatch(quorumToolOutput)
+	if matches == nil {
+		return false, errors.New("could not find Quorate line")
+	}
+
+	if string(matches[1]) == "Yes" {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func parseRings(cfgToolOutput []byte) []Ring {
+	re := regexp.MustCompile(`(?m)RING ID (?P<id>\d+)\s+id \s*= (?P<address>.+)\s+status \s*= (?P<status>.+)`)
+	matches := re.FindAllSubmatch(cfgToolOutput, -1)
+	rings := make([]Ring, len(matches))
+	for i, match := range matches {
+		namedMatches := extractRENamedCaptureGroups(re, match)
+
+		rings[i] = Ring{
+			Number:  namedMatches["id"],
+			Address: namedMatches["address"],
+			Faulty:  strings.Contains(namedMatches["status"], "FAULTY"),
+		}
+	}
+	return rings
+}
+
+func parseQuoromVotes(quorumToolOutput []byte) (quorumVotes QuorumVotes, err error) {
+	re := regexp.MustCompile(`(?m)Expected votes:\s+(\d+)\s+Highest expected:\s+(\d+)\s+Total votes:\s+(\d+)\s+Quorum:\s+(\d+)`)
+
+	matches := re.FindSubmatch(quorumToolOutput)
+	if matches == nil {
+		return quorumVotes, errors.New("could not find quorum votes numbers")
+	}
+
+	quorumVotes.ExpectedVotes, err = strconv.ParseUint(string(matches[1]), 10, 64)
```

Comment:
are we sure here we don't get a `panic: runtime error: index out of range` ?
In case match length is 0 only

Reply:
nope: since we have a mandatory capture group in the regex we will always have two matches in the result.
If it doesn't match at all, then we guard this in the two lines above.